src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreMetadata.cs

```diff
@@ -190,7 +190,7 @@ public bool Equals(ProjectRestoreMetadata other)
                    osStringComparer.Equals(OutputPath, other.OutputPath) &&
                    osStringComparer.Equals(ProjectName, other.ProjectName) &&
                    osStringComparer.Equals(ProjectUniqueName, other.ProjectUniqueName) &&
-                   Sources.OrderedEquals(other.Sources.Distinct(), source => source.Source, StringComparer.OrdinalIgnoreCase) &&
+                   Sources.Distinct().OrderedEquals(other.Sources.Distinct(), source => source.Source, StringComparer.OrdinalIgnoreCase) &&
```

Comment:
Many (most?) repos probably won't have duplicates, but doing `Distinct` twice is going to cause a bunch of additional allocations.  Would it be better to remove the `Distinct` instead, to get the (minor) perf benefit?

Reply:
NJ deduplicates at deserialization, while STJ does not. 
This is the only implementation that'll work for both.

Reply:
Won't that only affect people on the first restore after switching between NJ and STJ (including upgrading NuGet from a version that didn't previously support it).

Is avoiding one-time "warm" restore worth the every time perf cost?

Sometimes new NuGet features that modify the PackageSpec cause the DGSpec hash to change, so the first restore after a new NuGet version upgrade will fail the no-op check anyway. But it's only the first time after an upgrade, afterwards no-op still works.

Reply:
I'm not sure I follow the one time switch comments. 
The hashing is actually not affected, since we never hash a package spec from disk, it's only assets file dirty checks that are broken. 

I'll more details in the issue, but my intention is to switch to just use comparison without distinct once we've permanently switched to System.Text.Json.

Reply:
Ready for review.

Reply:
Where does NJ do the deduplication? Just curious where I missed this, the code where in JsonPackageSpecReader doesn't look it would dedup the sources.

Reply:
ReadObject seems to do it automatically. 
Couldn't figure out where/why, but it was just happening.